### PR TITLE
[TEVA-2315] Remove Hiring Staff Access to Application Data 1 Year After Closing Date

### PR DESCRIPTION
**Pull request by josephhull676**
## Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-2315

## Changes in this PR:

- Created a notification using the noticed gem. This notification, when delivered, creates a notification in the database and sends an email using the JobApplicationDataExpiryMailer
- Created a job to create/send the notification to all publishers associated with the hiring organisation when a vacancy is 2 weeks from the data expiry date (one year after the job listing ended)
- Added helper to remove link to view applications for vacancies that expired more than a year ago

## Screenshots of UI changes:

Screenshot 2021-06-16 at 17 00 45

### CODE REVIEW

File: app/helpers/notify_view_helper.rb [hunk 1]
```diff
@@ -79,6 +84,11 @@ def unsubscribe_link(subscription)
     notify_link(url, t(".unsubscribe_link_text"))
   end
 
+  def view_applications_for_link(vacancy)
```
Comment (david-mears-dfe) [inline app/helpers/notify_view_helper.rb:87]: We probably want to stick some tracking stuff in the url - see how the other urls in this helper use `**utm_params`. @stevenleggdfe can use this to track how many views to the page come from emails.

File: app/jobs/send_job_application_data_expiry_notification_job.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class SendJobApplicationDataExpiryNotificationJob < ApplicationJob
+  queue_as :default
+
+  def perform
+    start_of_today = DateTime.now.in_time_zone(Time.zone).beginning_of_day
+    end_of_today = DateTime.now.in_time_zone(Time.zone).end_of_day
+    two_weeks_less_than_a_year_ago = 1.year - 2.weeks
```
Comment (david-mears-dfe) [inline app/jobs/send_job_application_data_expiry_notification_job.rb:7]: `ago` is confusing and can be left out

File: app/jobs/send_job_application_data_expiry_notification_job.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    two_weeks_until_expiry = (start_of_today - two_weeks_less_than_a_year_ago)..(end_of_today - two_weeks_less_than_a_year_ago)
+
+    vacancies_with_two_weeks_until_expiry = Vacancy.expired.includes(organisations: :publishers).where(expires_at: two_weeks_until_expiry)
```
Comment (david-mears-dfe) [inline app/jobs/send_job_application_data_expiry_notification_job.rb:11]: for clear naming, can you do something like `vacancies_less_than_two_weeks_from_expiry`

File: spec/jobs/send_job_application_data_expiry_notification_job_spec.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+RSpec.describe SendJobApplicationDataExpiryNotificationJob do
+  subject(:job) { described_class.perform }
+
+  let(:organisation) { create(:school) }
+  let(:publisher) { create(:publisher) }
+  let(:two_weeks_less_than_a_year_ago) { Time.zone.now - (1.year - 2.weeks) }
+  let(:vacancy) { create(:vacancy, expires_at: two_weeks_less_than_a_year_ago, publisher: publisher, organisation_vacancies_attributes: [{ organisation: organisation }]) }
+
+  let(:notification) { double(:notification) }
```
Comment (david-mears-dfe) [inline spec/jobs/send_job_application_data_expiry_notification_job_spec.rb:11]: `instance_double` if you like. It complains when given a method not implemented by Notification.

File: app/helpers/notify_view_helper.rb (hunk 1 above)
Comment (stevenleggdfe) [reply app/helpers/notify_view_helper.rb:87]: Yes please! @josephhull676 if you follow the same scheme as the other emails use, email clickthrough rates will then be tracked automatically on the relevant dashboard.

File: spec/notifications/publishers/job_application_data_expiry_notification_spec.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      before do
+        travel_to 1.day.ago do
+          described_class
+            .with(vacancy: vacancy, publisher: publisher)
+            .deliver(publisher)
+        end
+      end
+
+      it "returns the correct timestamp" do
+        expect(subject).to match(/Yesterday at/)
```
Comment (alexbowen) [inline spec/notifications/publishers/job_application_data_expiry_notification_spec.rb:52]: is using a regex necessary here? (aassuming thats what this is - looks the same as JS)

Comment (josephhull676) [reply spec/notifications/publishers/job_application_data_expiry_notification_spec.rb:52]: The other notification spec checks against values given in the same way. I copied it lol

Comment (josephhull676) [reply spec/notifications/publishers/job_application_data_expiry_notification_spec.rb:52]: I guess it's only checking that the timestamp has "Yesterday at" in it. I'll change it now

Comment (josephhull676) [reply spec/notifications/publishers/job_application_data_expiry_notification_spec.rb:52]: I've changed these to use the include matcher instead now. Don't know why Connor used `to match` instead.

File: spec/jobs/send_job_application_data_expiry_notification_job_spec.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+require "rails_helper"
+
+RSpec.describe SendJobApplicationDataExpiryNotificationJob do
+  let(:notification) { instance_double(Publishers::JobApplicationDataExpiryNotification) }
+  let(:organisation) { create(:school) }
+  let(:publisher) { create(:publisher, organisation_publishers_attributes: [{ organisation: organisation }]) }
+  let!(:vacancy_with_two_weeks_to_data_expiry) { create(:vacancy, expires_at: 351.days.ago, publisher: publisher, organisation_vacancies_attributes: [{ organisation: organisation }]) }
+  let!(:vacancy) { create(:vacancy, expires_at: 1.day.ago, publisher: publisher, organisation_vacancies_attributes: [{ organisation: organisation }]) }
+
+  before do
```
Comment (cesidio) [inline spec/jobs/send_job_application_data_expiry_notification_job_spec.rb:10]: very minor thing, when is one liner I prefer it like this:

`before { allow(DisableExpensiveJobs).to receive(:enabled?).and_return(false) }`

**Review by cesidio — APPROVED** [review]
Nice work!

**Pull request merged**